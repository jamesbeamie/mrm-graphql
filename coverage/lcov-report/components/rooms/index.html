<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for components/rooms</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> components/rooms
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">0.72% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>3/416</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>0/121</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>0/55</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">0.74% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>3/403</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="AddRoomMenu.js"><a href="AddRoomMenu.js.html">AddRoomMenu.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	</tr>

<tr>
	<td class="file low" data-value="AddRoomNairobi.js"><a href="AddRoomNairobi.js.html">AddRoomNairobi.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="72" class="abs low">0/72</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="20" class="abs low">0/20</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="72" class="abs low">0/72</td>
	</tr>

<tr>
	<td class="file low" data-value="AddRoomToEpicTower.js"><a href="AddRoomToEpicTower.js.html">AddRoomToEpicTower.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="75" class="abs low">0/75</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="22" class="abs low">0/22</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="75" class="abs low">0/75</td>
	</tr>

<tr>
	<td class="file low" data-value="AddRoomToKampala.js"><a href="AddRoomToKampala.js.html">AddRoomToKampala.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="64" class="abs low">0/64</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="15" class="abs low">0/15</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="64" class="abs low">0/64</td>
	</tr>

<tr>
	<td class="file low" data-value="DeleteRoom.js"><a href="DeleteRoom.js.html">DeleteRoom.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="29" class="abs low">0/29</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="29" class="abs low">0/29</td>
	</tr>

<tr>
	<td class="file low" data-value="EditRoom.js"><a href="EditRoom.js.html">EditRoom.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="32" class="abs low">0/32</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="3" class="abs low">0/3</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="32" class="abs low">0/32</td>
	</tr>

<tr>
	<td class="file low" data-value="FilterRoomMenu.js"><a href="FilterRoomMenu.js.html">FilterRoomMenu.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="35" class="abs low">0/35</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="7" class="abs low">0/7</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="32" class="abs low">0/32</td>
	</tr>

<tr>
	<td class="file low" data-value="NairobiRoomInputs.js"><a href="NairobiRoomInputs.js.html">NairobiRoomInputs.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	</tr>

<tr>
	<td class="file low" data-value="Room.js"><a href="Room.js.html">Room.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="9" class="abs low">0/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="7" class="abs low">0/7</td>
	</tr>

<tr>
	<td class="file low" data-value="RoomForm.js"><a href="RoomForm.js.html">RoomForm.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	</tr>

<tr>
	<td class="file low" data-value="RoomSetup.js"><a href="RoomSetup.js.html">RoomSetup.js</a></td>
	<td data-value="4" class="pic low"><div class="chart"><div class="cover-fill" style="width: 4%;"></div><div class="cover-empty" style="width:96%;"></div></div></td>
	<td data-value="4" class="pct low">4%</td>
	<td data-value="25" class="abs low">1/25</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="19" class="abs low">0/19</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	<td data-value="4.35" class="pct low">4.35%</td>
	<td data-value="23" class="abs low">1/23</td>
	</tr>

<tr>
	<td class="file medium" data-value="Rooms.js"><a href="Rooms.js.html">Rooms.js</a></td>
	<td data-value="66.67" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 66%;"></div><div class="cover-empty" style="width:34%;"></div></div></td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="3" class="abs medium">2/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="3" class="abs medium">2/3</td>
	</tr>

<tr>
	<td class="file low" data-value="RoomsList.js"><a href="RoomsList.js.html">RoomsList.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="44" class="abs low">0/44</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="27" class="abs low">0/27</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="42" class="abs low">0/42</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Apr 11 2019 19:58:40 GMT+0300 (East Africa Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
