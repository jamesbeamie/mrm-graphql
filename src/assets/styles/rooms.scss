@import './variables';

.room-setup-card {
  overflow: auto;
  .room-details-container {
    .delete-room {
      position: relative;
      display: flex;
      float: right;
      margin-right: 2rem;
      justify-content: flex-end;
      button#modal-button.button {
        position: absolute;
        top: -2.3rem;
        border: none;
        font-size: 12px;
        visibility: hidden;
      }
      .edit-icon {
        position: absolute;
        top: -1.7rem;
        visibility: hidden;
      }
      h2 {
        text-transform: uppercase;
        font-size: 20px;
        font-weight: 700;
      }
      .delete-confirmation-text {
        color: $text-color;
        font-size: 18px;
        line-height: 1.7em;
      }

      .modal-component-delete {
        width: 100%;
        .modal_content .form-action-buttons {
          .btn-primary {
            margin-top: 35px;
            text-transform: uppercase;
          }
          .btn-secondary {
            margin-top: 35px;
            text-transform: uppercase;
          }
        }
      }
    }
  }
}

.room-setup-card:hover {
  overflow: auto;
  .room-details-container {
    .delete-room {
      justify-content: flex-end;
      .edit-icon {
        visibility: visible;
      }
      button#modal-button.button {
        visibility: visible;
        background: transparent;
      }
    }
  }
}
