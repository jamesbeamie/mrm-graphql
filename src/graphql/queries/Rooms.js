import gql from 'graphql-tag';
import { paginationFR, someRoomFR, roomFR, idNameFR } from './Fragments';

const remoteRoomFR = gql`
fragment remoteRoomFields on AllRemoteRooms {
  rooms {
    calendarId
    name
  }
}
`;

const roomDevicesFR = gql`
fragment roomDevices on Room {
  devices{
    ...idNameDevices
  }
}
${idNameFR('Devices')}`;

export const GET_ROOMS_QUERY = gql`
  query rooms($capacity: Int, $location: String!, $office: String!, $page: Int!, $perPage: Int!, $roomLabels: String) {
    allRooms(
      page: $page
      perPage: $perPage
      capacity: $capacity
      location: $location
      office: $office
      roomLabels: $roomLabels
    ) {
      rooms {
        ...roomFields
        ...roomDevices
      }
      ...pagination
    }
  }
${roomFR}
${roomDevicesFR}
${paginationFR('PaginatedRooms')}
`;

export const GET_ALL_ROOMS_QUERY = gql`
  query rooms($capacity: Int!, $location: String!, $office: String!, $page: Int, $perPage: Int) {
    allRooms(
      page: $page
      perPage: $perPage
      capacity: $capacity
      location: $location
      office: $office
    ) {
      rooms {
        ...roomFields
      }
    }
  }
${roomFR}
`;

export const GET_ALL_ROOMS = gql`
  query rooms($location: String){
    allRooms(location: $location) {
      rooms {
        ...roomFields
        ...roomDevices
      }
    }
  }
${roomFR}
${roomDevicesFR}`;

const GET_LOCATIONS_QUERY = gql`
  query locations {
    allLocations {
      ...idNameLocation
    }
  }
${idNameFR('Location')}`;

const GET_ROOM_BY_NAME = gql`
  query roomByName($name: String!) {
    getRoomByName(name: $name) {
      ...someRoomFields
      floor {
        block {
          offices {
            name
            location {
              ...idNameLocation
            }
          }
        }
      }
    }
  }
${someRoomFR}
${idNameFR('Location')}`;

const GET_ROOM = gql`
  query roomById($Id: Int) {
    getRoomById(roomId: $Id) {
      ...someRoomFields
      locationId
      roomLabels
      imageUrl
    }
  }
${someRoomFR}`;

const GET_ALL_REMOTE_ROOMS = gql`
  query {
    allRemoteRooms {
      ...remoteRoomFields
    }
  }
${remoteRoomFR}`;

const GET_REMOTE_ROOMS_ALL_LOCATIONS = gql`
  query allRooms($returnAll: Boolean) {
    allRemoteRooms(returnAll: $returnAll) {
      ...remoteRoomFields
    }
  }
${remoteRoomFR}`;

export {
  GET_ROOMS_QUERY as default,
  GET_LOCATIONS_QUERY,
  GET_ROOM_BY_NAME,
  GET_ALL_REMOTE_ROOMS,
  GET_REMOTE_ROOMS_ALL_LOCATIONS,
  GET_ROOM,
};
