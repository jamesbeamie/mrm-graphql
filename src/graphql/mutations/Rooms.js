import gql from 'graphql-tag';
import { roomFR, idNameFR } from '../queries/Fragments';

const EDIT_ROOM_DETAILS_MUTATION = gql`
  mutation updateRoom($roomId: Int!, $name: String!){
    updateRoom(roomId: $roomId,  name: $name){
      room{
        ...idNameRoom
      }
    }
  }
${idNameFR('Room')}`;

const DELETE_ROOM = gql`
  mutation DeleteRoom($roomId: Int!) {
    deleteRoom(roomId: $roomId) {
      room {
        ...idNameRoom
      }
    }
  }
${idNameFR('Room')}`;

const ADD_ROOM = gql`
  mutation AddRoom($name: String!, $roomType: String!, $capacity: Int!, $imageUrl: String!, $locationId: Int!, $calendarId: String!, $structureId: String!,
    $roomLabels: [String]){
    createRoom(name: $name, roomType: $roomType, capacity: $capacity, imageUrl: $imageUrl, locationId: $locationId, calendarId: $calendarId, structureId: $structureId
      roomLabels: $roomLabels){
        room {
          ...roomFields
        }
    }
  }
${roomFR}`;

export { EDIT_ROOM_DETAILS_MUTATION, DELETE_ROOM, ADD_ROOM };
