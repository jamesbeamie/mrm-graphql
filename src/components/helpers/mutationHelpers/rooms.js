import apolloClient from '../../../utils/ApolloClient';
import { ADD_ROOM, DELETE_ROOM } from '../../../graphql/mutations/Rooms';
import { GET_ROOMS_QUERY } from '../../../graphql/queries/Rooms';
import { getUserDetails } from '../../helpers/QueriesHelpers';

const addRoom = async (variables, location, client = apolloClient) => {
  const {
    imageUrl, name, capacity, roomType, calendarId, roomLabels, locationId, structureId,
  } = variables;
  await client.mutate({
    mutation: ADD_ROOM,
    name: 'addRoom',
    variables: {
      imageUrl, name, capacity, roomType, calendarId, roomLabels, locationId, structureId,
    },
    refetchQueries: [
      {
        query: GET_ROOMS_QUERY,
        variables: {
          location,
          office: '',
          page: 1,
          perPage: 8,
          roomLabels: '',
        },
      },
    ],

    update: async (proxy, { data: { createRoom: { room } } }) => {
      const cachedRooms = proxy.readQuery({
        query: GET_ROOMS_QUERY,
        variables: {
          location,
          office: '',
          page: 1,
          perPage: 8,
          roomLabels: '',
        },
      });
      const roomList = cachedRooms.allRooms.rooms;
      cachedRooms.allRooms.rooms = [...roomList, room];

      proxy.writeQuery({
        query: GET_ROOMS_QUERY,
        data: cachedRooms,
        variables: {
          location,
          office: '',
          page: 1,
          perPage: 8,
          roomLabels: '',
        },
      });
    },
  });
};

const getRoomsQueryObj = (location, theData = undefined) => (
  {
    query: GET_ROOMS_QUERY,
    variables: {
      location,
      office: '',
      page: 1,
      perPage: 8,
      roomLabels: '',
    },
    data: theData,
  }
);

const deleteRoom = async (roomId, client = apolloClient) => {
  const user = await getUserDetails();
  const roomsQueryObject = getRoomsQueryObj(user.location);
  let data;
  await client.mutate({
    mutation: DELETE_ROOM,
    name: 'deleteRoom',
    variables: {
      roomId,
    },
    refetchQueries: [roomsQueryObject],
    update: async (cache, { data: roomDelete }) => {
      data = cache.readQuery(roomsQueryObject);
      data.allRooms.rooms = data.allRooms.rooms.filter(
        room => room.id !== roomDelete.deleteRoom.room.id,
      );
      cache.writeQuery(getRoomsQueryObj(user.location, data));
    },
  });
  return data;
};

export {
  deleteRoom,
  addRoom,
};
