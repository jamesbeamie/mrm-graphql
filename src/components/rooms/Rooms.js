/* eslint-disable react/jsx-no-comment-textnodes */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable react/forbid-prop-types */
import React, { Component, createRef, Fragment } from 'react';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import ImageLoader from '../commons/ImageLoader';
import { deleteIcon, editIcon } from '../../utils/images/images';
import MrmModal from '../commons/MrmModal';
import '../../assets/styles/rooms.scss';
import '../../assets/styles/utility.scss';
import { deleteRoom } from '../helpers/mutationHelpers/rooms';
import notification from '../../utils/notification';
import EntityCard from './EntityCard';

class Room extends Component {
  state = {
    isLoading: false,
    error: '',
  };
  /**
   * It creates a reference to the modal element and Component
   * and toggle it to close when the close icon is clicked
   *
   */
  deleteRoomModal = createRef();

  toggleModal = () => {
    this.deleteRoomModal.current.toggleModal();
  };

  /**
   * It deletes the specified room from the list of rooms
   * and updates the UI to reflect the changes
   * @memberof Room
   */
  roomDelete = () => async () => {
    const { roomId, updatedRoom, roomName } = this.props;
    try {
      this.toggleLoading();
      const updatedRoomResponse = await deleteRoom(roomId);
      this.toggleLoading();
      this.toggleModal();
      notification(toastr, 'success', `${roomName} successfully deleted`)();
      this.props.clearFilter();
      updatedRoom(updatedRoomResponse);
    } catch (err) {
      this.toggleLoading();
      this.toggleModal();
      this.setState({
        error: 'Network error, kindly try again!',
      });
      notification(toastr, 'error', this.state.error)();
    }
  };

  /**
   * it change isLoading state to it's opposite value
   * i.e true to false or vise-versa
   *
   * @returns {void}
   */
  toggleLoading = () => {
    this.setState({
      isLoading: !this.state.isLoading,
    });
  };

  detailsView = () => {
    const {
      state: { showDetails, details },
    } = this;
    if (showDetails) {
      return (
        <EntityCard
          entityName={showDetails}
          entity={details}
          toggleEntity={this.toggleEntity}
        />
      );
    }
    // more if statements could be added to display details of more entities
    return undefined;
  };

  toggleEntity = (entityName, entity) => this.setState({
    showDetails: entityName,
    details: entity,
  });


  formatResources = (resources) => {
    const resource = resources.map(resourcesObject => resourcesObject.resource);
    return resource;
  };

  renderDeleteIcon = () => {
    const { roomName } = this.props;
    const { isLoading } = this.state;
    return (
      <MrmModal
        ref={this.deleteRoomModal}
        type={1}
        title="delete room"
        buttonText={
          <img src={deleteIcon} alt="Delete Icon" className="deletes-modal" />
        }
        modalContent={
          <div>
            <span className="delete-confirmation-text">
              Are you sure you want to delete &quot;{roomName}&quot;? This
              cannot be undone.
            </span>
          </div>
        }
        actionButtonText="DELETE"
        handleSubmit={this.roomDelete()}
        isLoading={isLoading}
      />
    );
  };

  renderEditIcon = () => (
    <span className="edit-icon">
      {/* // eslint-disable-next-line jsx-a11y/no-static-element-interactions */}
      <img
        className="img"
        src={editIcon}
        alt="edit-icon"
        onClick={() => this.props.handleClick(this.props)}
        onKeyPress={() => this.props.handleClick(this.props)}
      />
    </span>
  );

  renderRoomEntities = (entityName, entityCount, entity) => (
    <div className="number-of-resources">
      <button
        className="btn-plain negative-margin-Top-sm"
        onClick={() => this.toggleEntity(`${entityName}s`, entity)}
      >
        {`${entityCount || 'No'} ${entityName}${entityCount !== 1 ? 's' : ''}`}
      </button>
    </div>
  );

  renderRoomLabels = () => {
    const { roomLabels } = this.props;
    return (
      <div className="room-details">
        {roomLabels.map(label => (
          <div key={label} className="details">
            <p>{label}</p>
          </div>
        ))}
      </div>
    );
  };

  render() {
    const {
      props: {
        roomImage, roomName, numberOfSeats, resources, devices,
      },

      state: { showDetails },
    } = this;
    const resourceCount = resources.length;
    const deviceCount = devices ? devices.length : null;

    return (
      <div className="room-setup-card">
        {showDetails ? (
          this.detailsView()
        ) : (
          <Fragment>
            <ImageLoader
              source={roomImage}
              className="room-image"
              altText={roomName}
            />
            <div className="room-details-container">
              <div className="room-name">{roomName}</div>
              <div className="delete-room">
                {this.renderEditIcon()}
                {this.renderDeleteIcon()}
              </div>
              <div className="number-of-Seats">
                <p>Seats upto {numberOfSeats} people</p>
              </div>
              {this.renderRoomEntities('Resource', resourceCount, this.formatResources(resources))}
              {this.renderRoomEntities('Device', deviceCount, devices)}
              {this.renderRoomLabels()}
            </div>
          </Fragment>
          )}
      </div>
    );
  }
}

export default Room;

Room.propTypes = {
  roomImage: PropTypes.string.isRequired,
  roomName: PropTypes.string.isRequired,
  roomLabels: PropTypes.array.isRequired,
  numberOfSeats: PropTypes.number.isRequired,
  resources: PropTypes.instanceOf(Array).isRequired,
  roomId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  updatedRoom: PropTypes.func.isRequired,
  handleClick: PropTypes.func.isRequired,
  devices: PropTypes.instanceOf(Array),
  clearFilter: PropTypes.func.isRequired,
};

Room.defaultProps = {
  devices: [],
};
