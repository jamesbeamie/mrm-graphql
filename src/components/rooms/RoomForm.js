import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Input, SelectInput as Select, ActionButtons } from '../commons';
import { mapLocationsToSelectInputs } from '../../graphql/mappers/Rooms';

import '../../assets/styles/roomform.scss';

/**
 * Component for the room input form
 *
 * @returns {JSX}
 */
export class RoomForm extends Component {
  static propTypes = {
    roomLocation: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    onSubmit: PropTypes.func.isRequired,
    handleInputChange: PropTypes.func.isRequired,
    onCloseModalRequest: PropTypes.func.isRequired,
    formDetails: PropTypes.oneOfType([
      PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
          name: PropTypes.string,
        }),
      ),
      PropTypes.object,
    ]).isRequired,
    locations: PropTypes.arrayOf(PropTypes.object).isRequired,
    isLoading: PropTypes.bool,
  };

  static defaultProps = {
    roomLocation: '',
    isLoading: false,
  }

  /**
   * Submits the form data to the backend
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleSubmit = (event) => {
    event.preventDefault();
    this.props.onSubmit();
  };
  render() {
    const {
      onCloseModalRequest,
      locations,
      formDetails,
      roomLocation,
      handleInputChange,
      isLoading,
    } = this.props;
    return (
      <form onSubmit={this.handleSubmit}>
        <Input
          id="roomName"
          type="text"
          name="roomName"
          placeholder="Enter room name"
          value={formDetails.roomName}
          labelName="Room Name"
          onChange={handleInputChange}
          labelClass="input-wrapper"
          required
        />
        <Select
          labelText="Select Room Location"
          name="roomLocation"
          id="roomLocation"
          value={roomLocation}
          onChange={handleInputChange}
          key={roomLocation || ''}
          wrapperClassName="input-wrapper"
          placeholder="Select room location"
          options={mapLocationsToSelectInputs(locations)}
          disabled
        />
        <ActionButtons
          withCancel
          onClickCancel={onCloseModalRequest}
          isLoading={isLoading}
          onClickSubmit={this.handleSubmit}
        />
      </form>
    );
  }
}

export default RoomForm;
