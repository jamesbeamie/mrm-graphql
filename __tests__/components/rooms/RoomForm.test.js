import React from 'react';
import { mount } from 'enzyme';

import { RoomForm } from '../../../src/components/rooms/RoomForm';
import { roomLocations, formDetails } from '../../../__mocks__/rooms/Rooms';

describe('RoomForm', () => {
  let handleSubmit = jest.fn();
  const handleCloseModal = jest.fn();
  const wrapper = mount(<RoomForm
    onCloseModalRequest={handleCloseModal}
    onSubmit={handleSubmit}
    handleInputChange={jest.fn()}
    locations={roomLocations || ''}
    formDetails={formDetails}
  />);

  const event = {
    preventDefault: jest.fn(),
    target: {
      name: '',
      value: '',
    },
  };

  const wrapperValue = wrapper.find('RoomForm');
  it('should call handleSubmit', () => {
    const action = wrapperValue.instance();
    handleSubmit = jest.spyOn(wrapperValue.instance(), 'handleSubmit');
    action.handleSubmit(event);
    expect(handleSubmit).toBeCalled();
  });

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have one select and one text input element', () => {
    expect(wrapper.find('Input').length).toBe(1);
    expect(wrapper.find('SelectInput').length).toBe(1);
    expect(wrapper.find('ActionButtons').length).toBe(1);
  });
});
