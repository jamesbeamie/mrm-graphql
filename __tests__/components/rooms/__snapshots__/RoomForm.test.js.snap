// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomForm should render properly 1`] = `
<RoomForm
  formDetails={
    Object {
      "name": "HighLan",
      "roomId": "1",
    }
  }
  handleInputChange={[MockFunction]}
  isLoading={false}
  locations={
    Array [
      Object {
        "__typename": "Location",
        "id": "1",
        "name": "Kampala",
      },
      Object {
        "__typename": "Location",
        "id": "2",
        "name": "Lagos",
      },
      Object {
        "__typename": "Location",
        "id": "3",
        "name": "Nairobi",
      },
    ]
  }
  onCloseModalRequest={[MockFunction]}
  onSubmit={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
    }
  }
  roomLocation=""
>
  <form
    onSubmit={[Function]}
  >
    <Input
      controlsClass=""
      id="roomName"
      inputClass="mrm-input default-input"
      labelClass="input-wrapper"
      labelName="Room Name"
      name="roomName"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      placeholder="Enter room name"
      required={true}
      type="text"
      value="text"
    >
      <div
        className="input-wrapper"
      >
        <label
          htmlFor="roomName"
        >
          Room Name
          <input
            className="mrm-input default-input"
            defaultValue="text"
            id="roomName"
            name="roomName"
            onBlur={[Function]}
            onChange={[MockFunction]}
            onFocus={[Function]}
            placeholder="Enter room name"
            required={true}
            type="text"
          />
        </label>
      </div>
    </Input>
    <SelectInput
      disabled={true}
      id="roomLocation"
      isValue={false}
      key=""
      labelText="Select Room Location"
      name="roomLocation"
      onChange={[MockFunction]}
      onClick={[Function]}
      options={
        Array [
          Object {
            "id": "1",
            "name": "Kampala",
          },
          Object {
            "id": "2",
            "name": "Lagos",
          },
          Object {
            "id": "3",
            "name": "Nairobi",
          },
        ]
      }
      placeholder="Select room location"
      placeholderValue=""
      required={false}
      selectInputClassName="default-select"
      selected={-1}
      value=""
      wrapperClassName="input-wrapper"
    >
      <div
        className="input-wrapper"
      >
        <label
          htmlFor="roomLocation"
        >
          Select Room Location
          <form
            id="LocationFilters"
          >
            <select
              className="default-select"
              defaultValue=""
              disabled={true}
              id="roomLocation"
              name="roomLocation"
              onChange={[Function]}
              required={false}
              selected={-1}
            >
              <option
                value=""
              >
                Select room location
                 
              </option>
              <option
                key="Kampala"
                value="1"
              >
                Kampala
              </option>
              <option
                key="Lagos"
                value="2"
              >
                Lagos
              </option>
              <option
                key="Nairobi"
                value="3"
              >
                Nairobi
              </option>
            </select>
          </form>
        </label>
      </div>
    </SelectInput>
    <ActionButtons
      actionButtonText="SAVE CHANGES"
      cancelButtonText="CANCEL"
      isLoading={false}
      onClickCancel={[MockFunction]}
      onClickSubmit={[Function]}
      showActionButton={true}
      withCancel={true}
    >
      <div
        className="form-action-buttons"
      >
        <Button
          classProp={null}
          handleClick={[MockFunction]}
          isDisabled={false}
          title="CANCEL"
          type={1}
        >
          <button
            className="null btn-primary"
            disabled={false}
            onClick={[MockFunction]}
          >
            <span>
              CANCEL
            </span>
          </button>
        </Button>
        <Button
          classProp="main-button"
          handleClick={[Function]}
          isDisabled={false}
          title="SAVE CHANGES"
          type={2}
        >
          <button
            className="main-button btn-secondary"
            disabled={false}
            onClick={[Function]}
          >
            <span>
              SAVE CHANGES
            </span>
          </button>
        </Button>
      </div>
    </ActionButtons>
  </form>
</RoomForm>
`;
